File: BFS/542. 01 Matrix/Solution.java
// https://leetcode.com/problems/01-matrix/

/**
By Jiapeng
using BFS + mask
Runtime: 12 ms, faster than 82.98% of Java online submissions for 01 Matrix.
Memory Usage: 43.2 MB, less than 54.39% of Java online submissions for 01 Matrix.
**/
class Solution {
    /**
     step 1: create a full zero "map", all zeros in the map is set to 1
     step 2: use BFS, initialize the queue with all coordinate of zeros
     step 3: zeros begin to diffuse to its non-zero adjacent cells, update them with step 
     step 4: the cells that have already been checked should be added to "map" created in step 1
    **/
    public int[][] updateMatrix(int[][] matrix) {
        int xn = matrix.length; 
        int yn = matrix[0].length;
        if(xn == 0 || yn == 0)
            return matrix;
        
        int[][] map = new int[xn][yn];
        Queue<int[]> deque = new ArrayDeque<>();
        for(int i=0; i<xn; i++)
            for(int j=0; j<yn; j++){
                if(matrix[i][j] == 0){
                    deque.offer(new int[]{i, j});
                    map[i][j] = 1;
                }
            }
        
        // array for traverse adjacent cells
        int[] adjacent = {-1, 0, 1, 0, -1};
        int steps = 0;
        while(!deque.isEmpty()){
            int size = deque.size();
            while(size != 0){
                size--;
                int[] node = deque.poll();
                int row = node[0], col = node[1];
                // up, down, left and right
                for(int i=0; i<4; i++){
                    int x = row + adjacent[i];
                    int y = col + adjacent[i+1];
                    if(x < 0 || y < 0 || x >= xn || y >= yn || map[x][y] == 1)
                        continue;
                    deque.offer(new int[]{x, y});
                    map[x][y] = 1;
                    matrix[x][y] = steps + 1;
                }
            }
            steps++;
        }
        return matrix;
    }
}

File: BFS/542. 01 Matrix/Solution.py
# https://leetcode.com/problems/01-matrix/

# By Jiapeng
# BFS+mask, update matrix[x][y] with the variable steps
# Runtime: 812 ms, faster than 57.97% of Python3 online submissions for 01 Matrix.
# Memory Usage: 16.6 MB, less than 53.19% of Python3 online submissions for 01 Matrix.
class Solution:
    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
        """
        # step 1: create a full zero "map", all zeros in the map is set to 1
        # step 2: use BFS, initialize the queue with all coordinate of zeros
        # step 3: zeros begin to diffuse to its non-zero adjacent cells, update them with step 
        # step 4: the cells that have already been checked should be added to "map" created in step 1
        """
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return matrix
        mapp = [[0] * len(matrix[0]) for _ in range(len(matrix))]
        q = collections.deque()     # using deque can improve efficiency
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if matrix[i][j] == 0:
                    q.append((i, j))
                    mapp[i][j] = 1
        adjacent = [-1, 0, 1, 0, -1]
        steps = 0
        while q:
            size = len(q)
            while size:
                size -= 1
                node = q.popleft()
                for i in range(4):
                    x = node[0] + adjacent[i]
                    y = node[1] + adjacent[i+1]
                    if x < 0 or y < 0 or x >= len(matrix) or y >= len(matrix[0]) or mapp[x][y] == 1:
                        continue
                    q.append((x, y))
                    mapp[x][y] = 1
                    matrix[x][y] = steps + 1
            steps += 1
        return matrix
      
# By 
